### PR TITLE
Add MaxTime constrain property to has.

### PR DESCRIPTION
**Pull request by bmolnar2**
This change adds the feature of Has.MaxTime constraint for assertions: #2021

The implementation is based on https://gist.github.com/jnm2/f5c4e2224114d23e566467dbf1467128, however I rather organized out InvocationDescriptor from ThrowConstraint an resued that.

Please comment on the timeout values use in that test, whether they should be less or just fine with this timeout.

### CODE REVIEW

File: src/NUnitFramework/framework/Constraints/MaxTimeConstraint.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                    stopwatch.Start();
+                    object result = invocationDescriptor.Invoke();
+                    region.WaitForPendingOperationsToComplete(result);
+                    stopwatch.Stop();
+                }
+            }
+            else
+#endif
+            {
+                using (new TestExecutionContext.IsolatedContext())
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Constraints/MaxTimeConstraint.cs]: What's the purpose of using an IsolatedContext here? If the delegate should Assert, don't you want it to be reported?

File: src/NUnitFramework/framework/Constraints/Operators/MaxTimeOperator.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+using System;
+
+namespace NUnit.Framework.Constraints
+{
+    /// <summary>
+    /// Operator that tests if executed code terminated within
+    /// specified maximum time.
+    /// </summary>
+    public class MaxTimeOperator : SelfResolvingOperator
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Constraints/Operators/MaxTimeOperator.cs]: There appears to be know reason for this to be a SelfResolvingOperator. If there is a reason, please explain. In fact, I'm not clear why any operator at all is needed. This could be because I'm missing something in how this is supposed to be used. You don't actually have any tests that show it in use, only formal tests of the constraint.

File: src/NUnitFramework/framework/Constraints/MaxTimeConstraint.cs (hunk 1 above)
Comment (bmolnar2) [reply src/NUnitFramework/framework/Constraints/MaxTimeConstraint.cs]: Copy-paste error. It's not needed here, I'll remove that using.

**Comment by CharliePoole** [comment]
@rprouse I'm not convinced we want this feature at all and I think it's a shame that @bmolnar2 has done so much work on it. We should somehow continue and wrap up the discussion on the issue, which even now contains no solid examples of how this is supposed to be used.

@bmolnar2 I've only done the barest code review and asked some questions inline. IMO, it's a waste of time for you to actually make any changes until we resolve what we want to see. More on that on the issue.

**Comment by bmolnar2** [comment]
Ok, I'll stand down. Let me know if you need to revoke this one.

**Comment by jnm2** [comment]
Chronologically:

> I agree that we need something like that. If we all agree - as I suspect we do - then that moves us from the idea to the design stage

> I'm not convinced we want this feature at all

> On the substance of the feature: I think we all have agreed that there's a need for some more granular way to test the maximum time that things take within a test.

The middle statement seems harsh to me.

**Comment by CharliePoole** [comment]
@jnm2 Ref our separate discussion about chatter, you're generating notices again, in this case a rather cryptic one. Now I'm generating one to say "what's your point other than to criticise my language?"

IMO "I'm not convinced we want this feature" isn't harsh, it's a fairly dispassionate way of saying I didn't like "this feature" - whatever that was in the context.

Let's keep our discussions here substantive rather than ad hominem.


To that point, I'll re-summarize that I think we have agreed we need __some__ more granular feature but not on the syntax and precise semantics of it. We are collectively at the API design stage, not the coding stage, even though you may feel the design is settled.

**Comment by jnm2** [comment]
I apologize that my communication was cryptic and ad hominem. Not my intent at all, but I recognize that's the effect of it. My desire for visible communication and my desire to not go into a long meta came into conflict in that comment. I'll follow up with the concern I have via email.

**Comment by rprouse** [comment]
Closing because nobody has touched this PR in so long.
